### PR TITLE
Convert CMakeLists.txt to "pure cmake" version and add more cpp controllers

### PR DESCRIPTION
**Pull request by jviereck**
Port the catkin based CMakeLists.txt to "pure cmake" style. The python bindings are not working yet.

### CODE REVIEW

**Comment by jviereck** [comment]
This is the error I am getting when I try to enable the python bindings:

```
--- stderr: blmc_controllers                                 
CMake Warning at /home/jviereck_local/dev/learned_endeff_tracking/workspace/install/pybind11/share/cmake/pybind11/pybind11Tools.cmake:131 (add_library):
  Cannot generate a safe runtime search path for target py_blmc_controllers
  because files in some directories may conflict with libraries in implicit
  directories:

    runtime library [liburdfdom_sensor.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib
    runtime library [liburdfdom_model_state.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib
    runtime library [liburdfdom_model.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib
    runtime library [liburdfdom_world.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib

  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  CMakeLists.txt:71 (pybind11_add_module)


/usr/bin/ld: libblmc_controllers.a(impedance_controller.cpp.o): relocation R_X86_64_PC32 against symbol `_ZN9pinocchio8ModelTplIdLi0ENS_25JointCollectionDefaultTplEEC1Ev' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,7 +8,7 @@ project(blmc_controllers)
 if(NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
```
Comment (MaximilienNaveau) [inline CMakeLists.txt]: all this should be removed and replaced by the header you can find in here:
https://github.com/machines-in-motion/kino_dynamic_opt/blob/pure_cmake/momentumopt/CMakeLists.txt

File: CMakeLists.txt [hunk 2]
```diff
@@ -8,7 +8,7 @@ project(blmc_controllers)
...
   set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wl,--no-as-needed")
 endif()
-set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -W -Wextra")
+set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -W -Wextra -fPIC")
```
Comment (MaximilienNaveau) [inline CMakeLists.txt]: this can be removed all together, I do not think it buys you anything...

File: CMakeLists.txt [hunk 3]
```diff
@@ -20,70 +20,95 @@ set(CMAKE_INSTALL_RPATH ${CMAKE_INSTALL_PREFIX}/lib)
...
-)
+#
+# Dependencies
+#
+find_package(pybind11 CONFIG REQUIRED)
+find_package(Eigen3 REQUIRED)
+find_package(pinocchio REQUIRED)
+find_package(eiquadprog REQUIRED)
+SET(BOOST_COMPONENTS system python)
+find_package(Boost REQUIRED)
```
Comment (MaximilienNaveau) [inline CMakeLists.txt]: find_package(Boost REQUIRED COMPONENT python)

File: Config.cmake.in [hunk 4]
```diff
@@ -0,0 +1,11 @@
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/blmc_controllersTargets.cmake")
+
+include(CMakeFindDependencyMacro)
+
+# we do not add the other dependencies because these are header files lib
+find_dependency(yaml-cpp CONFIG REQUIRED)
+find_dependency(Eigen3 REQUIRED)
```
Comment (MaximilienNaveau) [inline Config.cmake.in:9]: wrong list, all these have target that needs to be found by the child project (basically all but the mpi_cmake_module):
```
find_dependency(pybind11 CONFIG REQUIRED)
find_dependency(Eigen3 REQUIRED)
find_dependency(pinocchio REQUIRED)
find_dependency(eiquadprog REQUIRED)
find_dependency(Boost REQUIRED COMPONENT python) # has a target inside that you need
find_dependency(yaml_utils REQUIRED) # has a target inside that you need
```

**Review by MaximilienNaveau — APPROVED** [review]

**Comment by MaximilienNaveau** [comment]
Some minor comments.

**Comment by MaximilienNaveau** [comment]
> /usr/bin/ld: libblmc_controllers.a(impedance_controller.cpp.o):

in your add library you missed the `SHARED` argument saying you want a `.so` (dynamic library) not a `.a` (static library)

File: CMakeLists.txt (hunk 3 above)
Comment (jviereck) [reply CMakeLists.txt]: I tried to change to `find_package(Boost REQUIRED COMPONENT python)` based on your suggestion, but then I am facing this error:

```
--- stderr: blmc_controllers
CMake Error at /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find Boost (missing: python) (found version "1.74.0")
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:577 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindBoost.cmake:2193 (find_package_handle_standard_args)
  CMakeLists.txt:29 (find_package)
```

Any thoughts?

File: CMakeLists.txt (hunk 1 above)
Comment (jviereck) [reply CMakeLists.txt]: I replaced it with a similar header as from the linked file.

File: CMakeLists.txt (hunk 2 above)
Comment (jviereck) [reply CMakeLists.txt]: This I took from the other header as well.

File: Config.cmake.in (hunk 4 above)
Comment (jviereck) [reply Config.cmake.in:9]: I've updated the `Config.cmake.in` based on your suggestion. For now I am using `SET(BOOST_COMPONENTS python)` still though as specifing the component on the find_dependency is not working (see other comment).

**Comment by jviereck** [comment]
> in your add library you missed the SHARED argument saying you want a .so (dynamic library) not a .a (static library)

Good catch! I've added the SHARED argument.

File: CMakeLists.txt (hunk 3 above)
Comment (MaximilienNaveau) [reply CMakeLists.txt]: `find_package(Boost REQUIRED COMPONENTS thread)` in CMakeLists.txt
and
`find_dependency(Boost REQUIRED COMPONENTS thread)` in Config.cmake.in

This works for the `solver` package.

Comment (MaximilienNaveau) [reply CMakeLists.txt]: Could it be that you are missing the `S` in `COMPONENTS`?

Comment (jviereck) [reply CMakeLists.txt]: You are right, the `S` was missing before.

According to `FindBoost` [1] page, it is required to specify the version of python to require (`python27` or `python36`)

When I try with `find_package(Boost REQUIRED COMPONENT python36)` I am getting:

```
CMake Error at /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find Boost (missing: python36) (found version "1.74.0")
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:577 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindBoost.cmake:2193 (find_package_handle_standard_args)
  CMakeLists.txt:29 (find_package)
```

When I try with `find_package(Boost REQUIRED COMPONENT python39)`, I am not getting an error on my machine anymore. So it looks like one has to specify the python version used by the system (on my mac python 3.9 is the default).

Do you have an idea how to figure out the current python version and then use it in the find boost command?

[1]: https://cmake.org/cmake/help/latest/module/FindBoost.html

Comment (MaximilienNaveau) [reply CMakeLists.txt]: have you tried the last comment of this answer: https://stackoverflow.com/questions/56067518/cmake-could-not-find-boost-python

Comment (MaximilienNaveau) [reply CMakeLists.txt]: If this works we can create a `get_python_version(python_version)` in mpi_cmake_module
Or just add these lines for all packages using boost python...

Comment (jviereck) [reply CMakeLists.txt]: Thanks for the link. This seems to work.

I've added the lines in the script directly as I assume there won't be much boost-python usage from our code.

I've added similar lines in the `Config.cmake.in` file. Is that okay to call `find_package(PythonLibs REQUIRED)` from there?

Comment (MaximilienNaveau) [reply CMakeLists.txt]: Yes it is ok I think...
Did you try to depend on this package?

Comment (jviereck) [reply CMakeLists.txt]: I added a dependency from the momentumopt package to this package. This seems to work (it compiles fine).

**Comment by MaximilienNaveau** [comment]
DO you use the branch from this PR https://github.com/machines-in-motion/mpi_cmake_modules/pull/27 ?

**Comment by jviereck** [comment]
yes

**Pull request merged**